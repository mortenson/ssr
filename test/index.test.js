/** @jsx h */

const { h } = require("@skatejs/val");
const render = require("..");

class CustomElement extends HTMLElement {
  constructor() {
    super();
    this.connectedCallback = jest.fn();
    this.disconnectedCallback = jest.fn();
  }
}
window.customElements.define("custom-element", CustomElement);

describe("window", () => {
  it("should work if re-included", () => {
    require("..");
  });

  it("Object", () => {
    expect(window.Object).toBeDefined();
    expect(window.Object).toEqual(global.Object);
  });

  describe("customElements", () => {
    it("should be an object", () => {
      expect(typeof window.customElements).toEqual("object");
    });

    it("define should add a the nodeName to the customElement", () => {
      expect(CustomElement.prototype.nodeName).toEqual("CUSTOM-ELEMENT");
    });

    it("get should return the custom element", () => {
      expect(window.customElements.get("custom-element")).toEqual(
        CustomElement
      );
    });
  });
});

describe("Node", () => {
  let host;

  beforeEach(() => {
    host = document.createElement("div");
  });

  describe("appendChild", () => {
    let m1, m2;

    it("should connect a custom element", () => {
      const node = document.createElement("custom-element");
      host.appendChild(node);
      expect(node.connectedCallback.mock.calls.length).toEqual(1);
    });

    it("should connect a fragment of elements", () => {});
  });

  describe("insertBefore", () => {
    it("should connect a custom element", () => {});

    it("should connect a fragment of elements", () => {});
  });

  describe("removeChild", () => {
    it("should connect a custom element", () => {});

    it("should connect a fragment of elements", () => {});
  });

  describe("replaceChild", () => {
    it("should connect a custom element", () => {});

    it("should connect a fragment of elements", () => {});
  });
});

describe("Element", () => {
  describe("attachShadow", () => {
    it("should be patched on Element", () => {
      expect(HTMLElement.prototype.hasOwnProperty("attachShadow")).toEqual(
        true
      );
    });

    it("mode: open, should create a shadowRoot property", () => {});

    it("mode: closed, should not create a shadowRoot property", () => {});
  });

  describe("innerHTML", () => {
    it("should get innerHTML", () => {});

    it("should set innerHTML", () => {
      const div = document.createElement("div");
      const html = `<h1 id="yelling">Test</h1><section><p><span id="nested-span">Paragraph</span> 1.</p><p><span>Paragraph</span> 2.</p></section>`;
      div.innerHTML = html;
      expect(div.innerHTML).toEqual(html);
    });
  });
});

describe("HTMLElement", () => {
  it("should extend Element", () => {
    expect(window.HTMLElement).toBeDefined();
  });
});

describe("render (serialisation)", () => {
  it("should call the connected callback", () => {
    const node = new CustomElement();
    return render(node).then(() => {
      expect(node.connectedCallback.mock.calls.length).toEqual(1);
    });
  });

  it("should disconnect it after serialising", () => {
    const node = new CustomElement();
    return render(node).then(() => {
      expect(node.disconnectedCallback.mock.calls.length).toEqual(1);
    });
  });

  describe("renders correctly", () => {
    class Hello extends HTMLElement {
      connectedCallback() {
        this.attachShadow({ mode: "open" }).appendChild(
          <span>
            Hello,{" "}
            <x-yell>
              <slot />
            </x-yell>!
            <x-yell>!</x-yell>
          </span>
        );
      }
    }
    class Yell extends HTMLElement {
      connectedCallback() {
        Promise.resolve().then(() => {
          this.attachShadow({ mode: "open" }).innerHTML = `
            <style>.test1,.test2 { font-weight: bold; }</style>
            <span class="test1"><slot></slot></span>
          `;
        });
      }
    }
    customElements.define("x-hello", Hello);
    customElements.define("x-yell", Yell);

    it("node", () => {
      const hello = new Hello();
      hello.appendChild(document.createTextNode("World"));

      return render(hello).then(r => {
        expect(r).toMatchSnapshot();
      });
    });

    it("document", () => {
      const hello = new Hello();
      hello.appendChild(document.createTextNode("World"));
      document.body.appendChild(hello);

      return render(document).then(r => {
        document.body.removeChild(hello);
        expect(r).toMatchSnapshot();
      });
    });
  });
});
