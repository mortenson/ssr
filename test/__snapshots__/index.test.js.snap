// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render (serialisation) renders correctly document 1`] = `
"<html><head><script>function __ssr_rehydrate(){var r=document.currentScript,e=r.parentNode,t=e.parentNode,n=function(r,e){for(;r&&r.firstChild;)e.appendChild(r.firstChild)};e.removeChild(r),t.removeChild(e);var o=t.attachShadow({mode:\\"open\\"});n(e,o);var a=o.querySelectorAll(\\"slot\\");(function(){for(var r,e=[],t=0,n=a.length;t<n;t++){var o=a[t];r&&r.contains(o)||(e.push(o),r=o)}return e})().forEach(function(r){n(r,t)})}function __ssr_restyle(){var t=document.currentScript,e=document.createElement(\\"style\\");e.textContent=document.getElementById(t.getAttribute(\\"data-style-id\\")).textContent,t.parentNode.replaceChild(e,t)}</script><style>.test10,.test20 { font-weight: bold; }</style></head><body><x-hello><shadow-root><span>Hello, <x-yell><shadow-root>
            <script data-style-id=\\"__style_0\\">__ssr_restyle()</script>
            <span class=\\"test10\\"><slot><slot>World</slot></slot></span>
          <script>__ssr_rehydrate()</script></shadow-root></x-yell>!<x-yell><shadow-root>
            <script data-style-id=\\"__style_0\\">__ssr_restyle()</script>
            <span class=\\"test10\\"><slot>!</slot></span>
          <script>__ssr_rehydrate()</script></shadow-root></x-yell></span><script>__ssr_rehydrate()</script></shadow-root></x-hello></body></html>"
`;

exports[`render (serialisation) renders correctly node 1`] = `
"<script>function __ssr_rehydrate(){var r=document.currentScript,e=r.parentNode,t=e.parentNode,n=function(r,e){for(;r&&r.firstChild;)e.appendChild(r.firstChild)};e.removeChild(r),t.removeChild(e);var o=t.attachShadow({mode:\\"open\\"});n(e,o);var a=o.querySelectorAll(\\"slot\\");(function(){for(var r,e=[],t=0,n=a.length;t<n;t++){var o=a[t];r&&r.contains(o)||(e.push(o),r=o)}return e})().forEach(function(r){n(r,t)})}function __ssr_restyle(){var t=document.currentScript,e=document.createElement(\\"style\\");e.textContent=document.getElementById(t.getAttribute(\\"data-style-id\\")).textContent,t.parentNode.replaceChild(e,t)}</script><style id=\\"__style_0\\">.test10,.test20 { font-weight: bold; }</style><x-hello><shadow-root><span>Hello, <x-yell><shadow-root>
            <script data-style-id=\\"__style_0\\">__ssr_restyle()</script>
            <span class=\\"test10\\"><slot><slot>World</slot></slot></span>
          <script>__ssr_rehydrate()</script></shadow-root></x-yell>!<x-yell><shadow-root>
            <script data-style-id=\\"__style_0\\">__ssr_restyle()</script>
            <span class=\\"test10\\"><slot>!</slot></span>
          <script>__ssr_rehydrate()</script></shadow-root></x-yell></span><script>__ssr_rehydrate()</script></shadow-root></x-hello>"
`;
